### PR TITLE
Mldb 1184 posneg

### PR DESCRIPTION
**Pull request by mailletf**

### CODE REVIEW

File: plugins/stats_table_procedure.cc [hunk 1]
```diff
@@ -543,6 +545,316 @@ run(const ProcedureRunConfig & run,
...
+                if(it == p_outcomes.end()) {
+                    return true;
+                }
+
+                rtnRow.push_back(make_tuple(Id(columnName.toUtf8String() + "_" + functionConfig.outcomeToUse),
+                                            it->second,
+                                            ts));
+
+
+//                 auto st = statsTables.find(columnName);
```
Comment (steadtler) [inline plugins/stats_table_procedure.cc]: Cleanup?

File: container_files/assets/www/doc/builtin/functions/BagOfWordsStatsTablePosNeg.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+## Input and Output Values
+
+Functions of this type have a single input value named `words` which is a row.
+
+Functions of this type have a single output value named `probs` which is also a row. 
+The row will contain the probability of the P(outcome | word) for each valid word 
+given the function configuration.
+
+## See also
+* The ![](%%doclink statsTable.bagOfWords.train procedure) trains a bag of words stats tables.
```
Comment (guyd) [inline container_files/assets/www/doc/builtin/functions/BagOfWordsStatsTablePosNeg.md]: tables -> table

File: plugins/stats_table_procedure.cc [hunk 3]
```diff
@@ -543,6 +545,316 @@ run(const ProcedureRunConfig & run,
...
+   
+    // sort all the keys by their p(outcome)
+    vector<pair<Utf8String, float>> accum;
+    for(auto it = statsTable.counts.begin(); it!=statsTable.counts.end(); it++) {
+        const pair<int64_t, std::vector<int64_t>> & counts = it->second;
+
+        if(counts.first < functionConfig.minTrials)
+            continue;
+
+        float poutcome = counts.second[outcomeToUseIdx] / float(counts.first);
```
Comment (steadtler) [inline plugins/stats_table_procedure.cc]: Could counts.first be zero / very small?

File: plugins/stats_table_procedure.cc [hunk 4]
```diff
@@ -543,6 +545,316 @@ run(const ProcedureRunConfig & run,
...
+        accum.push_back(make_pair(it->first, poutcome));
+    }
+
+    if(accum.size() < functionConfig.numPos + functionConfig.numNeg) {
+        for(const auto & col : accum)
+            p_outcomes.insert(col);
+    }
+    else {
+        auto comparer = [](const pair<Utf8String, float> & a,
```
Comment (steadtler) [inline plugins/stats_table_procedure.cc]: maybe rename comparer in english.

File: container_files/assets/www/doc/builtin/procedures/BagOfWordsStatsTableProcedure.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# Bag Of Words Stats Table Procedure
+
+This procedure type is meant to work with bags of words, like the output of the
```
Comment (guyd) [inline container_files/assets/www/doc/builtin/procedures/BagOfWordsStatsTableProcedure.md]: , like the output of -> as returned by

File: plugins/stats_table_procedure.cc [hunk 6]
```diff
@@ -543,6 +545,316 @@ run(const ProcedureRunConfig & run,
...
+    Date start = Date::now();
+
+    auto aggregator = [&] (const MatrixNamedRow & row,
+                           const std::vector<ExpressionValue> & extraVals)
+        {
+            if(num_req++ % 10000 == 0) {
+                double secs = Date::now().secondsSinceEpoch() - start.secondsSinceEpoch();
+                string progress = ML::format("done %d. %0.4f/sec", num_req, num_req / secs);
+                onProgress(progress);
+                cerr << progress << endl;
```
Comment (steadtler) [inline plugins/stats_table_procedure.cc]: remove?

File: plugins/stats_table_procedure.cc (hunk 3 above)
Comment (mailletf) [inline plugins/stats_table_procedure.cc]: An element would a count of zero would not be in the table

File: plugins/stats_table_procedure.cc (hunk 1 above)
Comment (mailletf) [inline plugins/stats_table_procedure.cc]: I'm going to be adding it back in a later PR

File: plugins/stats_table_procedure.cc [hunk 7]
```diff
@@ -156,7 +114,51 @@ reconstitute(ML::DB::Store_Reader & store)
...
 
 /*****************************************************************************/
-/* ROLL PROCEDURE                                                            */
+/* STATS TABLE PROCEDURE CONFIG                                              */
+/*****************************************************************************/
+
+DEFINE_STRUCTURE_DESCRIPTION(StatsTableProcedureConfig);
+
+StatsTableProcedureConfigDescription::
+StatsTableProcedureConfigDescription()
```
Comment (guyd) [inline plugins/stats_table_procedure.cc]: Were you planning to merge the sql expression into a single statement or will we do it latter?

Comment (mailletf) [inline plugins/stats_table_procedure.cc]: Later

File: plugins/stats_table_procedure.cc [hunk 8]
```diff
@@ -543,6 +545,316 @@ run(const ProcedureRunConfig & run,
...
+             "each row's counts will be influenced by rows that came before it.",
+             OrderByExpression::ROWHASH);
+    addField("outcomes", &BagOfWordsStatsTableProcedureConfig::outcomes,
+             "List of expressions to generate the outcomes. Each can be any expression "
+             "involving the columns in the dataset. The type of the outcomes "
+             "must be a boolean (0 or 1)");
+    addField("statsTableFileUrl", &BagOfWordsStatsTableProcedureConfig::statsTableFileUrl,
+             "URL where the stats table file (with extension '.st') should be saved. "
+             "This file can be loaded by a function of type 'statsTable.bagOfWords.posneg'.");
+}
```
Comment (guyd) [inline plugins/stats_table_procedure.cc]: If we want to support runOnCreation we must add the parent's field as well.

File: plugins/stats_table_procedure.cc [hunk 9]
```diff
@@ -156,7 +114,51 @@ reconstitute(ML::DB::Store_Reader & store)
...
+             "List of expressions to generate the outcomes. Each can be any expression "
+             "involving the columns in the dataset. The type of the outcomes "
+             "must be a boolean (0 or 1)");
+    addField("statsTableFileUrl", &StatsTableProcedureConfig::statsTableFileUrl,
+             "URL where the stats table file (with extension '.st') should be saved. "
+             "This file can be loaded by a function of type 'statsTable.getCounts'.");
+    addField("functionName", &StatsTableProcedureConfig::functionName,
+             "If specified, a 'statsTable.getCounts' function of this name will be "
+             "created using the trained stats tables.");
+}
```
Comment (guyd) [inline plugins/stats_table_procedure.cc]: If we want to support runOnCreation we should add the parent's field as well.

File: plugins/stats_table_procedure.cc [hunk 10]
```diff
@@ -543,6 +545,316 @@ run(const ProcedureRunConfig & run,
...
+               PolyConfig config,
+               const std::function<bool (const Json::Value &)> & onProgress)
+    : Function(owner)
+{
+    functionConfig = config.params.convert<StatsTablePosNegFunctionConfig>();
+
+    StatsTable statsTable;
+
+    // Load saved stats tables
+    ML::filter_istream stream(functionConfig.statsTableFileUrl.toString());
```
Comment (guyd) [inline plugins/stats_table_procedure.cc]: Will this throw a specific exception if the uri does not exist?

Comment (mailletf) [inline plugins/stats_table_procedure.cc]: Yup. Une erreur filter stream: 
   "contentType" : "application/json",
   "response" : "{\"error\":\"couldn't open file build/x86_64/tmp/mldb-873-stats_table_posneg.st.adsfasdf: No such file or directory\"}",
   "statusCode" : 400
}

**Comment by guyd** [comment]
+1

**Pull request merged**